apisix/plugins/jwt-auth.lua

```diff
@@ -76,7 +84,23 @@ local consumer_schema = {
                         },
                     },
                     required = {"public_key", "private_key"},
-                }
+                },
+                {
+                    properties = {
+                        vault = {
+                            type = "object",
+                            properties = {
+                                path = {type = "string"},
+                                add_prefix = {type = "boolean"}
```

Comment:
We should not allow users to specify path & add_prefix in the MVP, which will make thing very complex.
First, people may make mistakes, like using the same path according to multiple consumers.
Second, we need to enforce users to put their data under the same namespace, so we can provide a consistent token for APISIX to read the data.

Reply:
First:
Hi, thanks for the review. Without this feature, the MVP will be incomplete. Let me put down what it does:
suppose a user has the public key and private key stored into some arbitrary path `kv/rsa/token1`. It's very obvious users want to use it. What the plugin does is actually - if `path` is provided inside vault config it looks for fields like `public_key` and `private_key` (RS256) and `secret` (HS256 or HS512).
You can do it by
```shell
# the current directory contains public.pem and private.pem
vault kv put kv/rsa/token1 private_key=@private.pem public_key=@public.pem
```

PS: We are not storing consumer config into the vault so using the same path from multiple consumers won't be an issue as long as the user wants to do it. It just uses the secret from that path.

> people may make mistakes, like using the same path according to multiple consumers.

It's totally fine, for HS256 if the secret is not there in the auth_config and no data with key `secret` inside vault kv path - a new secret will be generated and be stored at that vault path. If someone reuses it, the stored secret inside the vault will be reused - there won't be any rewrite.

Second:
> Second, we need to enforce users to put their data under the same namespace, so we can provide a consistent token for APISIX to read the data.

Do you mean vault namespace? That's an enterprise feature [even I can't test it], we can look into it in the next version. WDYT?

Do let me know your views.

Reply:
There are two cases people what to set a common path (or ref):
1. they already have the data under another key
2. they want to share the same data among configurations

 For reason 1, since we already enforce the attributes name (secret, public key, .etc), it is not suitable to reuse the data format. And we may add new attribute requirements, which might conflict with the data already there.

For reason 2, it is quite strange to share the same configuration among consumers. Maybe we can add this feature when it's needed. (More features require more doc & test, and more bugs sometimes).

> Do you mean vault namespace? That's an enterprise feature [even I can't test it], we can look into it in the next version. WDYT?

Sorry for misleading you. Actually, I mean the keys can have a common prefix, like `/kv/apisix/`, which is easier to recognize and manage.

Reply:
> There are two cases people what to set a common path (or ref):
> 
> 1. they already have the data under another key
> 2. they want to share the same data among configurations
> 
> For reason 1, since we already enforce the attributes name (secret, public key, .etc), it is not suitable to reuse the data format.

One thing you are missing here, though we are enforcing key names (public_key) on the very first version [we can add user-provided key names in later versions]...we are letting them store the keys into the vault that can be referenced from the plugin at ease. Else the feature would be too much documentation oriented (they have to know from where the plugin will perform lookup and store it there). See the test case 14,15,16.

> And we may add new attribute requirements, which might conflict with the data already there.

Okay, we are only storing value in a vault kv path if secret is not provided in HS256/HS512. Maybe for the first version, we just read (if the secret is there, it uses it. if not it generate it and stores into plugin conf. but won't write anything), then users don't have to give us write access to the kv engine. Once we build up enough trust gradually, we can introduce edit capabilities into the plugin.
What are your views here?

> For reason 2, it is quite strange to share the same configuration among consumers. Maybe we can add this feature when it's needed. (More features require more doc & test, and more bugs sometimes).



> > Do you mean vault namespace? That's an enterprise feature [even I can't test it], we can look into it in the next version. WDYT?
> 
> Sorry for misleading you. Actually, I mean the keys can have a common prefix, like `/kv/apisix/`, which is easier to recognize and manage.

By allowing custom path or namespace we are not holding them back from storing key=value into kv/apisix/*  : )
To be honest, if I were the end user, I definitely want that flexibility instead of going through the docs and figuring out for jwt key name `foo` I have to store key into kv/apisix/jwt-auth/keys/foo.

Reply:
Btw, on thing I forgot to mention, both `path` and `add_prefix` are optional. Line 162

Reply:
Please remove the path & add_prefix, which is too flexible, and made future evolution harder. Let's be strict from the beginning. It's easier to relax the rule than be more strict with the rule. If people really need this feature, we can add it at that time.

Reply:
Done